### PR TITLE
Add F3DEX3 Microcode Option and Support

### PR DESCRIPTION
**Pull request by mountainflaw**
Version shipped is from 7b06c8117dc1748f7a1a855092c5dfa5109b55c5 on Fast3DEX3 master (https://github.com/HackerN64/F3DEX3/commit/7b06c8117dc1748f7a1a855092c5dfa5109b55c5).

Also fixes lighting direction and adds support for CameraWorld (required for fresnel and specular lighting).

### CODE REVIEW

File: Makefile
Comment (gheskett) [inline Makefile:105]: I'd personally leave this to use f3dzex by default for the time being, until HLE plugins are developed. Using f3dex3 should probably be a conscious choice as to not risk people that don't know what they're doing breaking emulators without a clue in the world of what's going on (discussion welcome idk).

File: Makefile
Comment (gheskett) [inline Makefile:123]: I'd get rid of this warning personally, _especially_ if f3dex3 is default. That should exist as a comment if at all.

File: src/game/rendering_graph_node.c
Comment (gheskett) [inline src/game/rendering_graph_node.c:655]: This shouldn't be changed unconditionally. Although, what is our HLE solution for this to make sure the game doesn't look wacky on most emulator plugins?

File: src/game/rendering_graph_node.c
Comment (gheskett) [inline src/game/rendering_graph_node.c:712]: Is this HLE safe?

**Review by gheskett — CHANGES_REQUESTED** [review]

File: src/game/rendering_graph_node.c
Comment (mountainflaw) [reply src/game/rendering_graph_node.c:712]: All new commands are noops for F3DEX2. Regardless, I don't think designing everything around faking the microcode in an emulator is a good idea.

File: Makefile
Comment (mountainflaw) [reply Makefile:105]: This is an accidental change. I will change it to F3DZEX.

File: src/game/rendering_graph_node.c
Comment (mountainflaw) [reply src/game/rendering_graph_node.c:655]: Those values are equivalent on standard microcode, since it's just the same thing but normalized. F3D/EX/EX2 just normalizes them behind the scenes, while EX3 does not.

File: Makefile
Comment (mountainflaw) [reply Makefile:123]: See above reply.

File: src/game/rendering_graph_node.c
Comment (gheskett) [reply src/game/rendering_graph_node.c:712]: I would agree, but people **_WILL_** attempt to play hacks on incompatible plugins no matter what.

Comment (mountainflaw) [reply src/game/rendering_graph_node.c:712]: If a hack creator is concerned about that, then F3DEX3 is not for them.

File: src/game/rendering_graph_node.c
Comment (gheskett) [reply src/game/rendering_graph_node.c:655]: Does this mean F3DEX2 still works with these values?

File: src/game/rendering_graph_node.c
Comment (mountainflaw) [reply src/game/rendering_graph_node.c:712]: > If a hack creator is concerned about that, then F3DEX3 is not for them.

To me at least, it's a moot point since you could say the same thing about libpl or libcart as well.

File: src/game/rendering_graph_node.c
Comment (mountainflaw) [reply src/game/rendering_graph_node.c:655]: Yes.

File: src/game/rendering_graph_node.c
Comment (gheskett) [reply src/game/rendering_graph_node.c:712]: libpl has checks in place to at least allow the user to design around both scenarios. Ideally, F3DEX3 should have some sort of runtime check that allows people to selectively do a thing depending on whether it's being used or not.

Comment (mountainflaw) [reply src/game/rendering_graph_node.c:712]: There is only a single check I can imagine would work, which would be reading the results of the built in F3DEX3 profiler. I'd expect F3DEX2 HLE to leave the data as zeros.

File: Makefile
Comment (gheskett) [reply Makefile]: Actually super risky apparently; will reluctantly accept keeping this in (as much as I like stderr being completely devoid of output)

Comment (mountainflaw) [reply Makefile]: Marking this as resolved since currently I can only get this working as a "warning".

File: Makefile
Comment (mountainflaw) [reply Makefile]: Corrected.

**Comment by mountainflaw** [comment]
Feedback addressed.

File: README.md
Comment (gheskett) [inline README.md:215]: Wait what even changed here? But tbh no please don't; the ultrasm64 stuff should probably be purged eventually but people should really be following the HackerSM64 wiki instead.

Comment (mountainflaw) [reply README.md:215]: oh I'm honestly not sure. This readme is nearly identical to the one for 2.2 minus the lines I added above.

Comment (mountainflaw) [reply README.md]: This random diff kinda disappeared somehow so this is resolved.

File: Makefile
Comment (gheskett) [inline Makefile:124]: Nit: These conditions should appear in the same order as the listed options above

File: src/boot/main.c
Comment (gheskett) [inline src/boot/main.c:76]: Nit: `#ifdef` is cleaner if it's a standalone condition

File: src/game/game_init.c
Comment (gheskett) [inline src/game/game_init.c:305]: These conditions should probably be checking `#elif defined()` since it's not the same thing as `#elifdef` (were that to be a real thing that exists)

File: src/s2d_engine/init.h
Comment (gheskett) [inline src/s2d_engine/init.h:21]: Ditto

**Review by gheskett — APPROVED** [review]

File: Makefile
Comment (gheskett) [reply Makefile:105]: Uncorrected

**Review by gheskett — CHANGES_REQUESTED** [review]

File: Makefile
Comment (gheskett) [reply Makefile]: Was also referring to the comments

File: Makefile
Comment (mountainflaw) [reply Makefile]: Corrected.

**Comment by 12abz** [comment]
good enough, added to my repo.

**Comment by mountainflaw** [comment]
This should be ready to merge! :)

**Comment by ThePerfectMario64** [comment]
merge this pr please

**Review by gheskett — DISMISSED** [review]

**Comment by mountainflaw** [comment]
Please do not merge until CameraWorld logic is fixed.

File: lib/PR/f3dex3/fifo/gspF3DEX3.debug.fifo.o
Comment (aglab2) [inline lib/PR/f3dex3/fifo/gspF3DEX3.debug.fifo.o:1]: I really dislike that .o files are posted in repo instead of doing similar approach to other microcodes. When F3DEX3 is built, we get data .code and .data files. Let's use them instead. A sample can be seen here https://github.com/aglab2/RoyalLegacy/commit/a6c6e433a13e71fd3a11e8d743a0859a85b57089

**Comment by aglab2** [comment]
Regarding backwards compatibility and RSP hang crash, it will never be resolved unless HLE plugins support new G_RELSEGMENT instruction https://github.com/HackerN64/F3DEX3/blob/main/gbi.h#L2616-L2619 . If this instruction is not supported, segments are not set causing aforementioned crash. Moreover even if that instruction is supported, lights loading is changed hence visuals will look weird. My opinion is that if dev wants to use F3DEX3, they have to accept that Jabo will break.

**Comment by ThePerfectMario64** [comment]
> Regarding backwards compatibility and RSP hang crash, it will never be resolved unless HLE plugins support new G_RELSEGMENT instruction https://github.com/HackerN64/F3DEX3/blob/main/gbi.h#L2616-L2619 . If this instruction is not supported, segments are not set causing aforementioned crash. Moreover even if that instruction is supported, lights loading is changed hence visuals will look weird. My opinion is that if dev wants to use F3DEX3, they have to accept that Jabo will break.

jabo sucks, so it's fine

**Comment by mountainflaw** [comment]
Succeeded by #816 .

**Pull request closed without merging**
